Goal: Make sure nmbd still responds to SIGTERM if it has no interfaces to
      listen on

Fixes: #168079

Status wrt upstream: forwarded to samba-technical 2007/05/30
                     will be in 3.0.26

Author: Steve Langasek <vorlon@debian.org>

Note: To be confirmed by Steve

Index: samba-3.0.25a/source/nmbd/nmbd_subnetdb.c
===================================================================
--- samba-3.0.25a.orig/source/nmbd/nmbd_subnetdb.c	2007-05-26 07:45:40.136219349 +0200
+++ samba-3.0.25a/source/nmbd/nmbd_subnetdb.c	2007-05-26 07:46:31.104625382 +0200
@@ -185,12 +185,16 @@
 	struct in_addr unicast_ip, ipzero;
 
 	if(num_interfaces == 0) {
+		void (*old_handler)(int);
+
 		DEBUG(0,("create_subnets: No local interfaces !\n"));
 		DEBUG(0,("create_subnets: Waiting for an interface to appear ...\n"));
+		old_handler = CatchSignal( SIGTERM, SIGNAL_CAST SIG_DFL );
 		while (iface_count() == 0) {
 			sleep(5);
 			load_interfaces();
 		}
+		CatchSignal( SIGTERM, SIGNAL_CAST old_handler );
 	}
 
 	num_interfaces = iface_count();
